File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第8章：边缘检测/8.4/scharr.py
# -*- coding: utf-8 -*-
import sys
import numpy as np
from scipy import signal
import cv2
def scharr(I,_boundary='symm'):
    # I 与 scharr_x 的 same 卷积
    scharr_x = np.array([[3,0,-3],[10,0,-10],[3,0,-3]],np.float32)
    I_x = signal.convolve2d(I,scharr_x,mode='same',boundary='symm')
    # I 与 scharr_y 的same 卷积
    scharr_y = np.array([[3,10,3],[0,0,0],[-3,-10,-3]],np.float32)
    I_y = signal.convolve2d(I,scharr_y,mode='same',boundary='symm')
    return (I_x,I_y)
#主函数
if __name__ =="__main__":
    if len(sys.argv)>1:
        image = cv2.imread(sys.argv[1],cv2.CV_LOAD_IMAGE_GRAYSCALE)
    else:
        print "Usge:python scharr.py imageFile"
    #求卷积
    i_conv_sch_x = scharr(image,1,0,_boundary='symm')
    i_conv_sch_y = scharr(image,0,1,_boundary='symm')
    #取绝对值,分别得到水平方向和垂直方向的边缘强度
    abs_i_conv_sch_x = np.abs(i_conv_sch_x)
    abs_i_conv_sch_y = np.abs(i_conv_sch_y)    
    #水平方向和垂直方向的边缘强度的灰度级显示
    edge_x = abs_i_conv_sch_x.copy()
    edge_y = abs_i_conv_sch_y.copy()
    edge_x[edge_x>255]=255
    edge_y[edge_y>255]=255
    edge_x = edge_x.astype(np.uint8)
    edge_y = edge_y.astype(np.uint8)
    cv2.imshow("edge_x",edge_x)
    cv2.imshow("edge_y",edge_y)
    cv2.imwrite("edge_x.jpg",edge_x)
    cv2.imwrite("edge_y.jpg",edge_y)
    #根据水平方向和垂直方向的边缘强度,求最终的边缘强度
    #有多种方式，这里使用平方根形式
    edge = np.sqrt(np.power(abs_i_conv_sch_x,2)+np.power(abs_i_conv_sch_y,2))
    #最终的边缘强度的灰度级显示
    edge[edge>255]=255
    edge = np.round(edge)
    edge = edge.astype(np.uint8)
    cv2.imshow('edge',edge)
    cv2.imwrite("edge.jpg",edge)
    #经过阈值处理的边缘显示
    cv2.namedWindow("thresh_edge",1)
    MAX_THRESH = 255
    thresh = 255
    #回调函数，阈值处理
    def callback_thresh(_thresh):
        threshEdge = edge.copy()
        threshEdge[threshEdge < _thresh] = 0
        threshEdge[threshEdge >= _thresh] = 255
        cv2.imshow("thresh_edge",threshEdge)
        cv2.imwrite("thresh_edge.jpg",threshEdge)
    callback_thresh(thresh)
    cv2.createTrackbar("thresh","thresh_edge",thresh,MAX_THRESH,callback_thresh)
    #模拟铅笔素描
    pencilSketch = edge.copy()
    pencilSketch = 255 - pencilSketch
    pencilSketch[pencilSketch < 100] = 100
    cv2.imshow("pencilSketch",pencilSketch)
    
    cv2.waitKey(0)
    cv2.destroyAllWindows()

File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第8章：边缘检测/8.4/scharr.cpp
#include<iostream>
using namespace std;
#include<opencv2/core/core.hpp>
#include<opencv2/imgproc/imgproc.hpp>
#include<opencv2/highgui/highgui.hpp>
using namespace cv;
Mat edge;//边缘图
int Thresh = 255;//阈值
const int MAX_THRESH = 255;
/*离散的二维卷积运算*/
void conv2D(InputArray _src, InputArray _kernel, OutputArray _dst, int ddepth, Point anchor = Point(-1, -1), int borderType = BORDER_DEFAULT)
{
	//卷积核顺时针旋转180
	Mat kernelFlip;
	flip(_kernel, kernelFlip, -1);
	//针对每一个像素,领域对应元素相乘然后相加
	filter2D(_src, _dst, CV_32FC1, kernelFlip, anchor, 0.0, borderType);
}
void scharr(InputArray src, OutputArray dst, int ddepth, int x, int y = 0, int borderType = BORDER_DEFAULT)
{
	CV_Assert(!(x == 0 && y == 0));
	Mat scharr_x = (Mat_<float>(3, 3) << 3, 0, -3, 10, 0, -10, 3, 0, -3);
	Mat scharr_y = (Mat_<float>(3, 3) << 3, 10, 3, 0, 0, 0, -3, -10, -3);
	//当 x 不等于零时，src 和 scharr_x 卷积
	if (x != 0)
	{
		conv2D(src, scharr_x, dst, ddepth, Point(-1, -1),borderType);
	}
	//当 y 不等于零时，src 和 scharr_y 卷积
	if (y != 0)
	{
		conv2D(src, scharr_y, dst, ddepth, Point(-1, -1), borderType);
	}
}
void callback_thresh(int, void*)
{
	Mat copyEdge = edge.clone();
	Mat thresh_edge;//阈值处理后的阈值
	threshold(copyEdge, thresh_edge, Thresh, MAX_THRESH, cv::THRESH_BINARY);
	imshow("阈值处理后的边缘强度", thresh_edge);
}
//主函数
int main(int argc, char*argv[])
{
	/*第一步：输入灰度图像矩阵*/
	Mat image = imread(argv[1], CV_LOAD_IMAGE_GRAYSCALE);
	if (!image.data)
	{
		cout << "没有图片" << endl;
		return -1;
	}
	/*第二步： scharr 卷积*/
	//图像矩阵和 scharr_x 卷积核卷积
	Mat img_scharr_x;
	scharr(image, img_scharr_x, CV_32FC1, 1, 0);
	//图像矩阵和 scharr_y 卷积核卷积
	Mat img_scharr_y;
	scharr(image, img_scharr_y, CV_32FC1, 0, 1);
	//两个卷积结果的灰度级显示
	Mat abs_img_scharr_x, abs_img_scharr_y;
	convertScaleAbs(img_scharr_x, abs_img_scharr_x, 1, 0);
	convertScaleAbs(img_scharr_y, abs_img_scharr_y, 1, 0);
	imshow("垂直方向的边缘", abs_img_scharr_x);
	imwrite("img1_sch_v_edge.jpg", abs_img_scharr_x);
	imshow("水平方向的边缘", abs_img_scharr_y);
	imwrite("img1_sch_h_edge.jpg", abs_img_scharr_y);
	/*第三步：通过第二步得到的两个卷积结果，求出最终的边缘强度*/
	//这里采用平方根的方式
	Mat img_scharr_x2, img_scharr_y2;
	pow(img_scharr_x, 2.0, img_scharr_x2);
	pow(img_scharr_y, 2.0, img_scharr_y2);
	sqrt(img_scharr_x2 + img_scharr_y2, edge);
	//数据类型转换，边缘强度的灰度级显示
	edge.convertTo(edge, CV_8UC1);
	imshow("边缘强度", edge);
	imwrite("img1_sch_edge.jpg", edge);
	//阈值处理后的边缘强度
	namedWindow("阈值处理后的边缘强度", cv::WINDOW_AUTOSIZE);
	createTrackbar("阈值", "阈值处理后的边缘强度", &Thresh, MAX_THRESH, callback_thresh);
	callback_thresh(0, 0);
	waitKey(0);
	return 0;
}